stan/math/rev/fun/diag_pre_multiply.hpp

```diff
@@ -0,0 +1,92 @@
+#ifndef STAN_MATH_REV_FUN_DIAG_PRE_MULTIPLY_HPP
+#define STAN_MATH_REV_FUN_DIAG_PRE_MULTIPLY_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/rev/core.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Return the product of the diagonal matrix formed from the vector
+ * or row_vector and a matrix.
+ *
+ * @tparam T1 type of the vector/row_vector
+ * @tparam T2 type of the matrix
+ * @param m1 input vector/row_vector
+ * @param m2 input matrix
+ *
+ * @return product of the diagonal matrix formed from the
+ * vector or row_vector and a matrix.
+ */
+template <typename T1, typename T2, require_eigen_vector_t<T1>* = nullptr,
+          require_eigen_t<T2>* = nullptr,
+          require_any_rev_matrix_t<T1, T2>* = nullptr>
+auto diag_pre_multiply(const T1& m1, const T2& m2) {
+  check_size_match("diag_pre_multiply", "m1.size()", m1.size(), "m2.rows()",
+                   m2.rows());
+  using inner_ret_type = decltype(value_of(m1).asDiagonal() * value_of(m2));
+  using ret_type = return_var_matrix_t<inner_ret_type, T1, T2>;
+  if (!is_constant<T1>::value && !is_constant<T2>::value) {
+    arena_t<promote_scalar_t<var, T1>> arena_m1 = m1;
+    arena_t<promote_scalar_t<var, T2>> arena_m2 = m2;
+    arena_t<ret_type> ret(arena_m1.val().asDiagonal() * arena_m2.val());
+    reverse_pass_callback([ret, arena_m1, arena_m2]() mutable {
+      arena_m1.adj().array() += arena_m2.val()
+                                    .array()
+                                    .cwiseProduct(ret.adj().array())
+                                    .rowwise()
+                                    .sum();
+      if (arena_m1.cols() >= arena_m1.rows()) {
+        arena_m2.adj().array() += arena_m1.val()
+                                      .array()
+                                      .transpose()
+                                      .replicate(1, arena_m2.cols())
+                                      .cwiseProduct(ret.adj().array());
+      } else {
+        arena_m2.adj().array() += arena_m1.val()
+                                      .array()
+                                      .replicate(1, arena_m2.cols())
+                                      .cwiseProduct(ret.adj().array());
+      }
+    });
+    return ret_type(ret);
+  } else if (!is_constant<T1>::value) {
+    arena_t<promote_scalar_t<var, T1>> arena_m1 = m1;
+    arena_t<promote_scalar_t<double, T2>> arena_m2 = value_of(m2);
+    arena_t<ret_type> ret(arena_m1.val().asDiagonal() * arena_m2);
+    reverse_pass_callback([ret, arena_m1, arena_m2]() mutable {
+      arena_m1.adj().array() += arena_m2.val()
+                                    .array()
+                                    .cwiseProduct(ret.adj().array())
+                                    .rowwise()
+                                    .sum();
+    });
+    return ret_type(ret);
```

Comment:
Instead of doing:

```cpp
arena_t<ret_type> ret(arena_m1.val().asDiagonal() * arena_m2);
reverse_pass_callback([ret, arena_m1, arena_m2]() mutable {
  arena_m1.adj().array() += arena_m2.val()
                                    .array()
                                    .cwiseProduct(ret.adj().array())
                                    .rowwise()
                                    .sum();
});
return ret_type(ret);
```

You can do:
```cpp
return make_callback_var(arena_m1.val().asDiagonal() * arena_m2,
    [arena_m1, arena_m2](const auto& ret_vi) mutable {
      arena_m1.adj() += arena_m2.val().asDiagonal() * ret_vi.adj();
});
```

It creates only one autodiff variable instead of two so is slightly more efficient (everything is packed into `make_callback_var` rather than a separate `ret` and `reverse_pass_callback`).

Reply:
@bbbales2 this does not seem to work for me. I get:

`inconsistent deduction for auto return type: 'Eigen::Matrix<stan::math::var_value<double>, 2, 2, 0, 2, 2>' and then 'stan::math::var_value<Eigen::Matrix<double, 2, 2>, void>'`

If I understand correctly, the `make_callback_var` returns a **varmat**. However, I was just talking about **varmat**s with Rok yesterday, and he told me that `diag_pre_multiply` does not support **varmat** yet. Could this be the reason, why this approach isn't working? Maybe I am way off, still learning a lot here.

Reply:
Oops, yeah you're right. I think the implementation you wrote is just a couple changes from working with varmat though. I will add a comment and if you want to varmat it you can, if not, that's fine.